IIS Hosting - Server Setup (#4826)
umbraco/UmbracoDocs: umbraco-cms/fundamentals/setup/server-setup/iis.md

---
description: "Information on hosting Umbraco on IIS"
---

# Hosting Umbraco in IIS

## Configuring IIS for .NET

> - Install [ASP.NET Core Runtime](https://dotnet.microsoft.com/en-us/download/dotnet/7.0) and download the **Hosting Bundle**.

Reviewer: ```suggestion
- Install the [ASP.NET Core Runtime](https://dotnet.microsoft.com/en-us/download/dotnet/7.0) and download the **Hosting Bundle**.
```

- Restart IIS (`net stop was /y` followed by `net start w3svc`)

> - Create a site in IIS-and ensure that the .NET Common Language Runtim (CLR) version` is set to `No Managed Code` for the Application Pool.

Reviewer: ```suggestion
- Create a site in IIS and ensure that the .NET Common Language Runtime (CLR) version is set to `No Managed Code` for the Application Pool.
```

dotnet publish -o ../deployment-artefacts -f net5.0
```

> Alternatively you can use File Transfer Protocal (FTP) publishing in Visual Studio to compile and collate all required files to for the application to run.

Reviewer: ```suggestion
Alternatively, you can use the File Transfer Protocol (FTP) publishing in Visual Studio to compile and collate all the required files for the application to run.
```

> In Visual Studio right, select the Umbraco web project in the *Solution Explorer* and choose the *Publish...* command.

Reviewer: ```suggestion
In Visual Studio, select the Umbraco web project in the *Solution Explorer* and choose the *Publish...* command.
```

![IIS Configuration Editor](images/iis-environment-variables.png)

Find the line named *environmentVariables* and open the dialog to add environment variables. These work similar to the *launchSettings*. You can define `ASPNETCORE_ENVIRONMENT` and create an `appSettings.[ASPNETCORE_ENVIRONMENT].json` file. Or even better create environment variables for sensitive settings like passwords. There are some differences to `launchSettings.json` configuration:

- Variable names need to change the object structure form JSON by combining the segments with double underscore `__` e.g. `ConnectionStrings__umbracoDbDSN`
- Escaped backslashes `\\` e.g. `serverName\\databaseInstanceName` are replaced by single backslash `\` e.g. `DATABASESERVER123\SQL2017`

### IIS Hosting models

IIS can host .NET applications using 2 different hosting models

- [In-process (default)](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/iis/in-process-hosting?view=aspnetcore-7.0)

> - In-process hosting runs an .NET app in the same process as its IIS worker process

Reviewer: ```suggestion
- In-process hosting runs a .NET app in the same process as its IIS worker process
```

- [Out-of-process](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/iis/out-of-process-hosting?view=aspnetcore-7.0) - to enable this model you need to edit your .csproj file and add:

```js
<PropertyGroup>
  <AspNetCoreHostingModel>OutOfProcess</AspNetCoreHostingModel>
</PropertyGroup>
```

> Out-of-process .NET apps run in a separate from the IIS worker process. The module controls the management of the Kestrel server and requests are proxied between them.

Reviewer: ```suggestion
Out-of-process .NET apps run separately from the IIS worker process. The module controls the management of the Kestrel server and requests are proxied between them.
```
